modules/js-and-css/optimize-oembeds/hooks.php

```diff
@@ -0,0 +1,115 @@
+}
+
+/**
+ * Filter the oEmbed HTML.
+ *
+ * Add loading="lazy" to any iframe tags.
+ * Lazy load any script tags.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $html The oEmbed HTML.
+ * @return string
+ */
+function perflab_optimize_oembed_html( $html ) {
+	$p = new WP_HTML_Tag_Processor( $html );
+
+	/**
+	 * Determine how to lazy load the embed.
+	 *
+	 * - If there is only one iframe, set loading="lazy".
+	 * - Prevent making scripts lazy if there is an inline script.
+	 *  - Only make script lazy if there is a single external script (since if there are
+	 *    multiple they may not get loaded in the right order).
+	 *  - Ensure that both the iframe and the script are made lazy if both occur in the same embed.
+	 */
+	$iframe_count      = 0;
+	$script_count      = 0;
+	$has_inline_script = false;
+	// Locate the iframes and scripts.
+	while ( $p->next_tag() ) {
+		if ( 'IFRAME' === $p->get_tag() ) {
+			$loading_value = $p->get_attribute( 'loading' );
+			if ( empty( $loading_value ) ) {
+				++$iframe_count;
+				$p->set_bookmark( 'iframe' );
+			}
+		} elseif ( 'SCRIPT' === $p->get_tag() ) {
+			if ( ! $p->get_attribute( 'src' ) ) {
+				$has_inline_script = true;
+			} else {
+				++$script_count;
+				$p->set_bookmark( 'script' );
+			}
+		}
+	}
+	// If there was only one non-inline script, make it lazy.
+	if ( 1 === $script_count && ! $has_inline_script ) {
+		add_action( 'wp_footer', 'perflab_optimize_oembed_lazy_load_scripts' );
+		$p->seek( 'script' );
+		$p->set_attribute( 'data-lazy-embed-src', $p->get_attribute( 'src' ) );
+		$p->remove_attribute( 'src' );
+	}
+	// If there was only one iframe, make it lazy.
+	if ( 1 === $iframe_count ) {
+		$p->seek( 'iframe' );
+		$p->set_attribute( 'loading', 'lazy' );
+	}
+	return $p->get_updated_html();
+}
+add_filter( 'embed_oembed_html', 'perflab_optimize_oembed_html' );
+
+/**
+ * Add a script to the footer if there are lazy loaded embeds.
+ * Load the embed's scripts when they approach the viewport using an IntersectionObserver.
+ *
+ * @since n.e.x.t
+ */
+function perflab_optimize_oembed_lazy_load_scripts() {
+	?>
+	<script type="module">
+		const lazyEmbedsScripts = document.querySelectorAll( 'script[data-lazy-embed-src]' );
+		const lazyEmbedScriptsByParents = new Map();
+
+		const lazyEmbedObserver = new IntersectionObserver(
+			( entries ) => {
+				for ( const entry of entries ) {
+					if ( entry.isIntersecting ) {
+						const lazyEmbedParent = entry.target;
+						const lazyEmbedScript = lazyEmbedScriptsByParents.get( lazyEmbedParent );
+						const embedScript = document.createElement( 'script' );
+						for ( const attr of lazyEmbedScript.attributes ) {
+							embedScript.setAttribute(
+								attr.nodeName === 'data-lazy-embed-src' ? 'src' : attr.nodeName,
+								attr.nodeValue
+							);
+						}
+						lazyEmbedScript.replaceWith( embedScript );
+						lazyEmbedObserver.unobserve( lazyEmbedParent );
+					}
+				}
+			},
+			{
+				rootMargin: '0px 0px 500px 0px',
```

Comment:
Curious how the `500px` value was determined? I wonder if it should rather be something like `1vh`?

I recall AMP has different values for this, depending on how heavy the component is. For example, `amp-youtube` uses [`0.75vh`](https://github.com/ampproject/amphtml/blob/ef8eac730183fba961e15b77d61b4ff7ccb36c6d/extensions/amp-youtube/0.1/amp-youtube.js#L129-L136).

[By default](https://github.com/ampproject/amphtml/blob/ef8eac730183fba961e15b77d61b4ff7ccb36c6d/src/base-element.js#L506-L518) AMP would render a lazy-loaded element when scrolling within three viewports (`3vh`) of the element, for example `amp-img`.

Reply:
Ah good question, this was largely experimental. I believe Chrome native lazy loading is even more conservative with a value over 1000px. It is also sensitive to low bandwidth, raising the threshold. 

I like the viewport suggestion, maybe `1vh`. Do you think we should add a filter here to make it adjustable?

Reply:
Also seeing another IntersectionObserver in AMP which [uses](https://github.com/ampproject/amphtml/blob/5df810730e3fc56f119e8ee5812bf2699ac7dde8/extensions/amp-list/1.0/component.js#L124) percentages here: 

```
bufferPct = Math.floor(viewportBuffer * 100)
...
rootMargin: `0% 0% ${bufferPct}% 0%`,
```

[Also](https://github.com/ampproject/amphtml/blob/5df810730e3fc56f119e8ee5812bf2699ac7dde8/extensions/amp-a4a/0.1/within-viewport.js#L61-L64):

```js
  const observer = new win.IntersectionObserver(callback, {
    root,
    rootMargin: `${(viewportNum - 1) * 100}%`,
  });
```

I guess `1vh` would do the same? It's not immediately clear if the percentage refers to the viewport or the element size.

There's also the two patterns for `rootMargin`, of `0 0 X 0` and just `X` (i.e. `X X X X`). The latter would make sense when scrolling toward an embed in any direction other than from top to bottom, such as when a page is loaded with an anchor link I think.

As for a filter, maybe? But maybe later if there is a clear need.

Reply:
How about this:

```suggestion
				rootMargin: '100% 0% 100% 0%',
```

(I previously used `1vh` when I meant `100vh` as the unit is divided into 100.)

This will cause the element to be initialized when scrolling the element into view from the top _or_ bottom when within 1 viewport.

Reply:
thanks for researching, accepting your suggestion for now - we can always fine tune later